File: servers/scalatra/src/main/scala/apis/TradeApi.scala
package apis

import com.wordnik.client.model.Trade
import com.wordnik.client.model.TradeBin
import java.io.File

import org.scalatra.{ TypedParamSupport, ScalatraServlet }
import org.scalatra.swagger._
import org.json4s._
import org.json4s.JsonDSL._
import org.scalatra.json.{ JValueResult, JacksonJsonSupport }
import org.scalatra.servlet.{FileUploadSupport, MultipartConfig, SizeConstraintExceededException}

import scala.collection.JavaConverters._

class TradeApi (implicit val swagger: Swagger) extends ScalatraServlet 
    with FileUploadSupport
    with JacksonJsonSupport
    with SwaggerSupport {
  protected implicit val jsonFormats: Formats = DefaultFormats

  protected val applicationDescription: String = "TradeApi"
  override protected val applicationName: Option[String] = Some("trade")

  before() {
    contentType = formats("json")
    response.headers += ("Access-Control-Allow-Origin" -> "*")
  }



  val getOperation = (apiOperation[List[Trade]]("get")
      summary "Get Trades."
      parameters(
        queryParam[String]("symbol").description(""),queryParam[Any]("filter").description(""),queryParam[List[String]]("columns").description(""),queryParam[Double]("start").description(""),queryParam[Boolean]("reverse").description(""),queryParam[Date]("startTime").description(""),queryParam[Date]("endTime").description(""),queryParam[Double]("count").description("").defaultValue(100))
  )

  get("/",operation(getOperation)) {
    val symbol = params.getAs[String]("symbol")
    println("symbol: " + symbol)
  val filter = params.getAs[Any]("filter")
    println("filter: " + filter)
  val columns = params.getAs[List[String]]("columns")
    println("columns: " + columns)
  val start = params.getAs[Double]("start")
    println("start: " + start)
  val reverse = params.getAs[Boolean]("reverse")
    println("reverse: " + reverse)
  val startTime = params.getAs[Date]("startTime")
    println("startTime: " + startTime)
  val endTime = params.getAs[Date]("endTime")
    println("endTime: " + endTime)
  val count = params.getAs[Double]("count")
    println("count: " + count)
  }




  val getBucketedOperation = (apiOperation[List[TradeBin]]("getBucketed")
      summary "Get previous trades in time buckets."
      parameters(
        queryParam[String]("symbol").description(""),queryParam[Any]("filter").description(""),queryParam[List[String]]("columns").description(""),queryParam[Double]("start").description(""),queryParam[Boolean]("reverse").description(""),queryParam[Date]("startTime").description(""),queryParam[Date]("endTime").description(""),queryParam[String]("binSize").description("").defaultValue("1m"),queryParam[Double]("count").description("").defaultValue(100))
  )

  get("/bucketed",operation(getBucketedOperation)) {
    val symbol = params.getAs[String]("symbol")
    println("symbol: " + symbol)
  val filter = params.getAs[Any]("filter")
    println("filter: " + filter)
  val columns = params.getAs[List[String]]("columns")
    println("columns: " + columns)
  val start = params.getAs[Double]("start")
    println("start: " + start)
  val reverse = params.getAs[Boolean]("reverse")
    println("reverse: " + reverse)
  val startTime = params.getAs[Date]("startTime")
    println("startTime: " + startTime)
  val endTime = params.getAs[Date]("endTime")
    println("endTime: " + endTime)
  val binSize = params.getAs[String]("binSize")
    println("binSize: " + binSize)
  val count = params.getAs[Double]("count")
    println("count: " + count)
  }




  val getByDateOperation = (apiOperation[List[Trade]]("getByDate")
      summary "Get trades between two dates. [Deprecated, use GET /trades]"
      parameters(
        queryParam[String]("symbol").description(""),queryParam[Date]("startTime").description(""),queryParam[Date]("endTime").description(""))
  )

  get("/byDate",operation(getByDateOperation)) {
    val symbol = params.getAs[String]("symbol")
    println("symbol: " + symbol)
  val startTime = params.getAs[Date]("startTime")
    println("startTime: " + startTime)
  val endTime = params.getAs[Date]("endTime")
    println("endTime: " + endTime)
  }




  val getRecentOperation = (apiOperation[List[Trade]]("getRecent")
      summary "Get recent trades. [Deprecated, use GET /trades]"
      parameters(
        queryParam[String]("symbol").description(""),queryParam[Double]("count").description("").defaultValue(100))
  )

  get("/recent",operation(getRecentOperation)) {
    val symbol = params.getAs[String]("symbol")
    println("symbol: " + symbol)
  val count = params.getAs[Double]("count")
    println("count: " + count)
  }

}

File: clients/ruby/lib/trade_api.rb
require "uri"

class Trade_api
  basePath = "https://www.bitmex.com/api/v1"
  # apiInvoker = APIInvoker

  def self.escapeString(string)
    URI.encode(string.to_s)
  end

  def self.get (symbol,filter,columns,start,reverse,start_time,end_time,count= 100,opts={})
    query_param_keys = [:symbol,:filter,:columns,:count,:start,:reverse,:start_time,:end_time]

    # set default values and merge with input
    options = {
    :symbol => symbol,
      :filter => filter,
      :columns => columns,
      :start => start,
      :reverse => reverse,
      :start_time => start_time,
      :end_time => end_time,
      :count => count}.merge(opts)

    #resource path
    path = "/trade".sub('{format}','json')

    
    # pull querystring keys from options
    queryopts = options.select do |key,value|
      query_param_keys.include? key
    end
    
    headers = nil
    post_body = nil
    response = Swagger::Request.new(:GET, path, {:params=>queryopts,:headers=>headers, :body=>post_body }).make.body
    response.map {|response|Trade.new(response)}

  end

def self.get_bucketed (symbol,filter,columns,start,reverse,start_time,end_time,bin_size= "1m",count= 100,opts={})
    query_param_keys = [:bin_size,:symbol,:filter,:columns,:count,:start,:reverse,:start_time,:end_time]

    # set default values and merge with input
    options = {
    :symbol => symbol,
      :filter => filter,
      :columns => columns,
      :start => start,
      :reverse => reverse,
      :start_time => start_time,
      :end_time => end_time,
      :bin_size => bin_size,
      :count => count}.merge(opts)

    #resource path
    path = "/trade/bucketed".sub('{format}','json')

    
    # pull querystring keys from options
    queryopts = options.select do |key,value|
      query_param_keys.include? key
    end
    
    headers = nil
    post_body = nil
    response = Swagger::Request.new(:GET, path, {:params=>queryopts,:headers=>headers, :body=>post_body }).make.body
    response.map {|response|TradeBin.new(response)}

  end

def self.get_by_date (symbol,start_time,end_time,opts={})
    query_param_keys = [:symbol,:start_time,:end_time]

    # verify existence of params
    raise "start_time is required" if start_time.nil?
    # set default values and merge with input
    options = {
    :symbol => symbol,
      :start_time => start_time,
      :end_time => end_time}.merge(opts)

    #resource path
    path = "/trade/byDate".sub('{format}','json')

    
    # pull querystring keys from options
    queryopts = options.select do |key,value|
      query_param_keys.include? key
    end
    
    headers = nil
    post_body = nil
    response = Swagger::Request.new(:GET, path, {:params=>queryopts,:headers=>headers, :body=>post_body }).make.body
    response.map {|response|Trade.new(response)}

  end

def self.get_recent (symbol,count= 100,opts={})
    query_param_keys = [:symbol,:count]

    # verify existence of params
    raise "count is required" if count.nil?
    # set default values and merge with input
    options = {
    :symbol => symbol,
      :count => count}.merge(opts)

    #resource path
    path = "/trade/recent".sub('{format}','json')

    
    # pull querystring keys from options
    queryopts = options.select do |key,value|
      query_param_keys.include? key
    end
    
    headers = nil
    post_body = nil
    response = Swagger::Request.new(:GET, path, {:params=>queryopts,:headers=>headers, :body=>post_body }).make.body
    response.map {|response|Trade.new(response)}

  end

end

